File: backend/app/core/cuisines.py
"""Cuisine taxonomy — the backend mirror of frontend `lib/cuisines.ts`.

Two-level ids: a top-level `"indian"` or a child `"indian/gujarati"`. The
frontend sends these ids in `RecommendRequest.cuisines`; a recipe carries a
normalized `cuisine` (top level) + optional `region` (child). Matching is OR
across the selected ids; a top-level id matches any region under it.

Asian and Indian are separate top-level groups (per the user's questionnaire).
"""
from __future__ import annotations

from typing import Optional

CUISINE_TAXONOMY: dict[str, list[str]] = {
    "asian": ["chinese", "thai", "japanese", "filipino", "korean", "vietnamese"],
    # Indian regions expanded once the real corpus populated them
    # (each of these has 80+ recipes). The importer already stores these slugs.
    "indian": ["north_indian", "south_indian", "punjabi", "gujarati", "marathi",
               "bengali", "kerala", "tamil_nadu", "karnataka", "rajasthani",
               "andhra", "goan"],
    "italian": [],
    "mexican": [],
    "mediterranean": [],
    "middle-eastern": [],
    "american": [],
}

# Human labels for every id (top-level + "top/child"). Source of truth for the
# GET /v1/cuisines response; the frontend renders whatever the endpoint returns.
CUISINE_LABELS: dict[str, str] = {
    "asian": "Asian", "asian/chinese": "Chinese", "asian/thai": "Thai",
    "asian/japanese": "Japanese", "asian/filipino": "Filipino",
    "asian/korean": "Korean", "asian/vietnamese": "Vietnamese",
    "indian": "Indian", "indian/north_indian": "North Indian",
    "indian/south_indian": "South Indian", "indian/punjabi": "Punjabi",
    "indian/gujarati": "Gujarati", "indian/marathi": "Marathi",
    "indian/bengali": "Bengali", "indian/kerala": "Kerala",
    "indian/tamil_nadu": "Tamil Nadu", "indian/karnataka": "Karnataka",
    "indian/rajasthani": "Rajasthani", "indian/andhra": "Andhra",
    "indian/goan": "Goan", "italian": "Italian", "mexican": "Mexican",
    "mediterranean": "Mediterranean", "middle-eastern": "Middle Eastern",
    "american": "American",
}


def label_for(cid: str) -> str:
    """Human label for a cuisine id, with a titled fallback for unknown slugs."""
    return CUISINE_LABELS.get(cid, cid.split("/")[-1].replace("_", " ").title())

# All valid ids: "asian", "asian/chinese", "indian/gujarati", ...
VALID_CUISINE_IDS: set[str] = {
    top for top in CUISINE_TAXONOMY
} | {
    f"{top}/{child}" for top, children in CUISINE_TAXONOMY.items() for child in children
}


def parse_cuisine_id(cid: str) -> tuple[str, Optional[str]]:
    """`"indian/gujarati"` -> ("indian", "gujarati"); `"italian"` -> ("italian", None)."""
    top, _, region = cid.partition("/")
    return top, (region or None)


def cuisine_matches(
    recipe_cuisine: Optional[str],
    recipe_region: Optional[str],
    selected_ids: list[str],
) -> bool:
    """True if the recipe satisfies ANY selected cuisine id (OR semantics)."""
    if not selected_ids:
        return True
    if not recipe_cuisine:
        return False
    for cid in selected_ids:
        top, region = parse_cuisine_id(cid)
        if recipe_cuisine != top:
            continue
        if region is None or recipe_region == region:
            return True
    return False

File: frontend/lib/cuisines.ts
// Cuisine taxonomy — mirror of backend/app/core/cuisines.py CUISINE_TAXONOMY.
// Asian and Indian are separate top-level groups; the questionnaire drills into
// a group's common cuisines with a "Doesn't matter" escape that sends the parent
// id ("indian"), while specific picks send child ids ("indian/gujarati").

export interface CuisineNode {
  id: string;
  label: string;
  count?: number;          // live recipe count (from GET /v1/cuisines); absent in the fallback
  children?: CuisineNode[];
}

export const CUISINES: CuisineNode[] = [
  {
    id: "asian",
    label: "Asian",
    children: [
      { id: "asian/chinese", label: "Chinese" },
      { id: "asian/thai", label: "Thai" },
      { id: "asian/japanese", label: "Japanese" },
      { id: "asian/filipino", label: "Filipino" },
      { id: "asian/korean", label: "Korean" },
      { id: "asian/vietnamese", label: "Vietnamese" },
      { id: "asian/sri_lankan", label: "Sri Lankan" },
      { id: "asian/malaysian", label: "Malaysian" },
      { id: "asian/indonesian", label: "Indonesian" },
      { id: "asian/nepalese", label: "Nepalese" },
      { id: "asian/burmese", label: "Burmese" },
    ],
  },
  {
    id: "indian",
    label: "Indian",
    children: [
      { id: "indian/north_indian", label: "North Indian" },
      { id: "indian/south_indian", label: "South Indian" },
      { id: "indian/punjabi", label: "Punjabi" },
      { id: "indian/gujarati", label: "Gujarati" },
      { id: "indian/marathi", label: "Marathi" },
      { id: "indian/bengali", label: "Bengali" },
      { id: "indian/kerala", label: "Kerala" },
      { id: "indian/tamil_nadu", label: "Tamil Nadu" },
      { id: "indian/karnataka", label: "Karnataka" },
      { id: "indian/rajasthani", label: "Rajasthani" },
      { id: "indian/andhra", label: "Andhra" },
      { id: "indian/goan", label: "Goan" },
    ],
  },
  { id: "italian", label: "Italian" },
  { id: "mexican", label: "Mexican" },
  { id: "mediterranean", label: "Mediterranean" },
  { id: "middle-eastern", label: "Middle Eastern" },
  { id: "american", label: "American" },
  {
    id: "european",
    label: "European",
    children: [
      { id: "european/british", label: "British" },
      { id: "european/irish", label: "Irish" },
      { id: "european/french", label: "French" },
      { id: "european/spanish", label: "Spanish" },
      { id: "european/portuguese", label: "Portuguese" },
      { id: "european/dutch", label: "Dutch" },
      { id: "european/german", label: "German" },
      { id: "european/polish", label: "Polish" },
      { id: "european/russian", label: "Russian" },
      { id: "european/ukrainian", label: "Ukrainian" },
      { id: "european/croatian", label: "Croatian" },
      { id: "european/slovak", label: "Slovak" },
      { id: "european/norwegian", label: "Norwegian" },
    ],
  },
  {
    id: "latin-american",
    label: "Latin American",
    children: [
      { id: "latin-american/brazilian", label: "Brazilian" },
      { id: "latin-american/colombian", label: "Colombian" },
      { id: "latin-american/argentine", label: "Argentine" },
      { id: "latin-american/venezuelan", label: "Venezuelan" },
      { id: "latin-american/uruguayan", label: "Uruguayan" },
      { id: "latin-american/peruvian", label: "Peruvian" },
      { id: "latin-american/chilean", label: "Chilean" },
      { id: "latin-american/cuban", label: "Cuban" },
    ],
  },
  {
    id: "caribbean",
    label: "Caribbean",
    children: [
      { id: "caribbean/jamaican", label: "Jamaican" },
      { id: "caribbean/trinidadian", label: "Trinidadian" },
      { id: "caribbean/haitian", label: "Haitian" },
    ],
  },
  {
    id: "african",
    label: "African",
    children: [
      { id: "african/algerian", label: "Algerian" },
      { id: "african/kenyan", label: "Kenyan" },
      { id: "african/nigerian", label: "Nigerian" },
      { id: "african/ethiopian", label: "Ethiopian" },
      { id: "african/south_african", label: "South African" },
    ],
  },
];
